File: Python3/2874_maximum_value_of_an_ordered_triplet_ii.py
# https://leetcode.cn/problems/maximum-value-of-an-ordered-triplet-ii/

class Solution:
    def maximumTripletValue(self, nums):
        """
        Inputs:
            nums: list[int]
        Outputs:
            res: int
        """
        n = len(nums)
        right_max = [0 for _ in range(n)]
        for i in range(n-2, -1, -1):
            right_max[i] = max(right_max[i+1], nums[i+1])
            
        res, left_max = 0, nums[0]
        for i in range(1, n-1):
            res = max(res, (left_max - nums[i]) * right_max[i])
            left_max = max(left_max, nums[i])
        return res * (res > 0)

File: Cpp/2874_maximum_value_of_an_ordered_triplet_ii.cpp
// https://leetcode.cn/problems/maximum-value-of-an-ordered-triplet-ii/

#include <iostream>
#include <vector>

using namespace std;
        
class Solution {
public:
    long long maximumTripletValue(vector<int>& nums) {
        int n = nums.size();
        vector<int> right_max(n, 0);
        for (int i = n-2; i >= 0; --i) {
            right_max[i] = max(right_max[i+1], nums[i+1]);
        }
        long long res = 0, left_max = nums[0];
        for (int i = 1; i < n-1; ++i) {
            res = max(res, (left_max - nums[i]) * right_max[i]);
            left_max = max(left_max, (long long)nums[i]);
        }
        return res > 0 ? res : 0;
    }
};

int main() {
    Solution sol;
    vector<int> nums = {12,6,1,2,7};
    cout << sol.maximumTripletValue(nums) << '\n';
    return 0;
}